### PR TITLE
feat(app-shell): Add configuration file with CLI and env overrides

### PR DESCRIPTION
**Pull request by mcous**
## overview

This PR implements the electron side of app configuration as described in #1463 in order to support #1329.

## changelog

- feat(app-shell): Add configuration file with CLI and env overrides

## review requests

There's +182 of README diff in this one and +50 of package.json + lockfile, so the actual code diff isn't that large.

For testing this, I've been using `ui.width` and `ui.height` because it's readily apparent if they work. But also, the `make dev` task in `app-shell` has been updated to use the CLI args so there's some testing built in.

 - [x] Config from file is read (test by modifying `config.json`, location described in `app-shell/README.md`)
- [x] Config from env var overrides file
    - e.g. `make dev OT_APP_UI__WIDTH=768`
- [x] Config from CLI overrides file
    - Makefile overrides console log level in `make dev` to `debug` (default is `info`)
- [x] Config from CLI overrides env var
    - e.g. Verify that `make dev OT_APP_LOG__LEVEL__CONSOLE=error` has no effect because of the CLI arg in the Makefile

Also please double check my documentation in `README.md` (thanks @IanLondon for reminding me to do this).

### CODE REVIEW

File: app-shell/lib/config.js [hunk 1]
```diff
@@ -1,17 +1,15 @@
...
-const url = require('url')
+const Store = require('electron-store')
+const dotProp = require('dot-prop')
+const mergeOptions = require('merge-options')
+const yargsParser = require('yargs-parser')
 
-// TODO(mc, 2018-05-15): "properly" means features, not DEV_MODE / DEBUG
-const DEV_MODE = process.env.NODE_ENV === 'development'
-const DEBUG_MODE = process.env.DEBUG
+const ENV_PREFIX = 'OT_APP'
```
Comment (mcous) [inline app-shell/lib/config.js]: How does everyone feel about this env. var. prefix?

File: app-shell/lib/config.js [hunk 2]
```diff
@@ -21,12 +19,50 @@ module.exports = {
...
+    webPreferences: {
+      webSecurity: true
+    }
+  }
+}
+
+const overrides = yargsParser(process.argv.slice(2), {
+  envPrefix: ENV_PREFIX,
+  configuration: {
+    'negation-prefix': 'disable_'
```
Comment (mcous) [inline app-shell/lib/config.js]: How does everyone feel about this custom negation prefix? Default for `yargs-parser` is `no-`, and I could just let the default be what we use

```shell
# set boolean "nested.foo" to true and "nested.bar" to false
# our custom config
./Opentrons --nested.foo --disable_nested.bar

# default yargs-parser
./Opentrons --nested.foo --no-nested.bar
--
```

The reason this negation prefix matters is that `yargs-parser` is a little complicated with boolean params. While it _does_ parse numbers, `--something=true` and `--something=false` (and `OT_APP_SOMETHING=true` + `OT_APP_SOMETHING=false`) are parsed as the strings `"true"` and `"false"` (both of which are truthy in JS) unless each boolean parameter is explicitly called out in the yargs-parser options.

If explicitly calling out each boolean parameter seems better than this whole negation prefix mess, we can also just do that.

File: yarn.lock [hunk 3]
```diff
@@ -2807,6 +2817,12 @@ dot-prop@^4.1.0, dot-prop@^4.1.1:
   dependencies:
     is-obj "^1.0.0"
 
+dot-prop@^4.2.0:
```
Comment (IanLondon) [inline yarn.lock]: When I run `yarn` on this branch, it removes this section and merges dot-prop 4.2.0 with the previous dot-prop section. Not sure why that's happening, does `yarn` produce no diff on your machine?

Comment (mcous) [reply yarn.lock]: Looks like I missed a lockfile commit

**Pull request merged**